ladybug_geometry/polyskel.py

```diff
@@ -0,0 +1,694 @@
+
+class _LAV:
+	def __init__(self, slav):
+		self.head = None
+		self._slav = slav
+		self._len = 0
+		log.debug('Created LAV %s', self)
+
+
+	@classmethod
+	def from_polygon(cls, polygon, slav):
+		lav = cls(slav)
+		for prev, point, next in _window(polygon):
+			lav._len += 1
+			vertex = _LAVertex(
+				point,
+				LineSegment2D.from_end_points(prev, point),
+				LineSegment2D.from_end_points(point, next)
+				)
+
+			vertex.lav = lav
+			if lav.head == None:
+				lav.head = vertex
+				vertex.prev = vertex.next = vertex
+			else:
+				vertex.next = lav.head
+				vertex.prev = lav.head.prev
+				vertex.prev.next = vertex
+				lav.head.prev = vertex
+		return lav
+
+	@classmethod
+	def from_chain(cls, head, slav):
+		lav = cls(slav)
+		lav.head = head
+		for vertex in lav:
+			lav._len += 1
+			vertex.lav = lav
+		return lav
+
+	def invalidate(self, vertex):
+		assert vertex.lav is self, 'Tried to invalidate a vertex thats not mine'
+		log.debug('Invalidating %s', vertex)
+		vertex._valid = False
+		if self.head == vertex:
+			self.head = self.head.next
+		vertex.lav = None
+
+	def unify(self, vertex_a, vertex_b, point):
+		normed_b_bisector = vertex_b.bisector.v.normalize()
+		normed_a_bisector = vertex_a.bisector.v.normalize()
+		replacement =_LAVertex(
+			point,
+			vertex_a.edge_left,
+			vertex_b.edge_right,
+		 	(normed_b_bisector, normed_a_bisector)
+			)
+		replacement.lav = self
+
+		if self.head in [vertex_a, vertex_b]:
+			self.head = replacement
+
+		vertex_a.prev.next = replacement
+		vertex_b.next.prev = replacement
+		replacement.prev = vertex_a.prev
+		replacement.next = vertex_b.next
+
+		vertex_a.invalidate()
+		vertex_b.invalidate()
+
+		self._len -= 1
+		return replacement
+
+	def __str__(self):
+		return 'LAV {}'.format(id(self))
+
+	def __repr__(self):
+		return '{} = {}'.format(str(self), [vertex for vertex in self])
+
+	def __len__(self):
+		return self._len
+
+	def __iter__(self):
+		cur = self.head
+		while True:
+			yield cur
+			cur = cur.next
+			if cur == self.head:
+				raise StopIteration
+
+	def _show(self):
+		cur = self.head
+		while True:
+			print(cur.__repr__())
+			cur = cur.next
+			if cur == self.head:
+				break
+
+class _EventQueue:
+	def __init__(self):
+		self.__data = []
+
+	def put(self, item):
+		if item is not None:
+			heapq.heappush(self.__data, item)
+
+	def put_all(self, iterable):
+		for item in iterable:
+			heapq.heappush(self.__data, item)
+
+	def get(self):
+		return heapq.heappop(self.__data)
+
+	def empty(self):
+		return len(self.__data)==0
+
+	def peek(self):
+		return self.__data[0]
+
+	def show(self):
+		for item in self.__data:
+			print(item)
+
+
+def skeletonize(polygon, holes=None):
+	"""
+	Compute the straight skeleton of a polygon.
+
+	The polygon should be given as a list of vertices in counter-clockwise order.
+	Holes is a list of the contours of the holes, the vertices of which should be in
+	clockwise order.
```

Comment:
I'll clarify. It'll be a list of tuples with each tuple representing a vertice. Holes and polygons are the same, with the exception that the former is in cw order and polygons are in ccw order.

Reply:
Ok. Maybe a code sample here would also help clear up any confusion.

On an unrelated and less-important note, the singular of "vertices" is usually "vertex" and I saw that you had some comments in your code that called it a "vertice". It's not that important as we still know what you mean and English is incredibly inconsistent in its rules. But it might be nice to always call it a "vertex" for our own sake of consistency.